Integrations  testing functionality

arivepr opened the pull request:
Addressing [the following Jira Card](https://issues.redhat.com/browse/RHCLOUD-25193)

Adding test functionality to integrations appearing in the integrations table, using internal pop-up notifications.

<img width="908" alt="Screenshot 2023-11-28 at 11 54 50 AM" src="https://github.com/RedHatInsights/notifications-frontend/assets/56621323/e2a83fa4-b617-4711-a46f-fe17f5484294">

<img width="920" alt="Screenshot 2023-11-28 at 11 55 32 AM" src="https://github.com/RedHatInsights/notifications-frontend/assets/56621323/41c23d3e-faeb-418c-8a70-5fee09daa62b">

![D815251C-D1A6-44D5-84EA-6BD6AA58DF14](https://github.com/RedHatInsights/notifications-frontend/assets/56621323/3d46e5c7-9eb4-4b65-a75f-cb89fbba871a)


As per previous discussion, leaving the Feature flag commented out for now given it forces

arivepr committed:
First pass on test functionality

Debugging test modal

arivepr committed:
Implementing test functionality, debugging pop ups.

arivepr commented:
Just a quick update: Currently exploring wrapping the component with the notificationRegistry. Had not realized we're using registry on this repo, trying to see if by explicitly wrapping this component specifically, the Pop ups will then start rendering properly.

Re-update: Did not fix it :D

arivepr committed:
Wrapping test modal with Provider

arivepr committed:
Functioning pop-up notifications after server response.

arivepr committed:
Modal now reopening consistently.

Final cleanup

Readding preview flag

Testing new logic

Reverting accidental changes

florkbr commented on src/components/Integrations/Table.tsx:
```diff
@@ -351,6 +351,7 @@ export const IntegrationsTable: React.FunctionComponent<IntegrationsTableProps>
     const actionsResolverCallback: IActionsResolver = React.useCallback(
       (rowData) => {
         const actionResolver = props.actionResolver;
+
```
@arivepr nit - while I know this _looks_ more sensible than it did with the extra new line - I usually try to avoid checking in these changes with other code as it could cause merge conflicts with someone else working on this file (especially if there are no other changes to the file in your PR).

florkbr commented on src/pages/Integrations/List/List.tsx:
```diff
@@ -191,10 +202,12 @@ const IntegrationsList: React.FunctionComponent<IntegrationListProps> = ({
     onEnable: integrationRows.onEnable,
   });
 
+  // eslint-disable-next-line
```
@arivepr is this necessary? It looks like this line was passing linter before?

florkbr commented on src/components/Integrations/Table/IntegrationTestProvider.tsx:
nit and not blocking - I wish we used another word internally to refer to this feature other than `Test` as upon initial glance I thought this was a test file and not production code/functionality.

florkbr commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,71 @@
+  const { addSuccessNotification, addWarningNotification } = useNotification();
+
+  const placeholderText =
+    'Congratulations! The integration you created on https://console.redhat.com was successfully tested!';
```
@arivepr hardcoding `https://console.redhat.com` may not be ideal as this will also appear on `https://console.stage.redhat.com`?

florkbr replied:
Also this seems a bit odd as placeholder text to me. It almost makes it feel like the integration is tested _before_ you submit the request. Did this come from UX?

InsaneZein commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,71 @@
+  const failedTestText =
+    'Your integration test has unfortunately failed, please verify your integration information.';
+
+  const handleModalCancel = () => {
```
Why wrap a function in a function? It doesn't seem necessary.

arivepr replied:
Ah nice catch!  Pushed logic to the parent, never got rid of the original func lol

I'll clean it up in a sec.

InsaneZein reviewed (approved):
just one lil comment. Looks good to me otherwise.

arivepr committed:
Cleaning up closing functionality.

Had moved logic up the parent chain, never got rid of the original function and instead just added the new one. Clearing that out.

Fixing error

florkbr commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,67 @@
+    'Congratulations! The integration you created on https://console.redhat.com was successfully tested!';
+
+  const failedTestText =
+    'Your integration test has unfortunately failed, please verify your integration information.';
```
@arivepr where did we get these strings from?

florkbr commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,67 @@
+    'Your integration test has unfortunately failed, please verify your integration information.';
+
+  const handleNotificationTest = async (notificationMessage: string) => {
+    const endpointURL = `/api/integrations/v1/endpoints/${integrationUUID}/test`;
```
Not unique to this PR - but we should try to move to a common route builder in case the root API version or path ever changes.

florkbr commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,67 @@
+            'Failed Test',
+            `Error through server response: ${response.status}`
+          );
+      onClose();
```
@arivepr since `onClose();` is executed in both the `try` and `catch`, should we move it to `finally` instead and dedupe? Also I'd imagine this conditional logic is hit when the server returns 400 or 500 level errors. Is showing a notification with just the `response.status` actually enough information?

florkbr commented on src/pages/Integrations/List/List.tsx:
```diff
@@ -278,7 +295,14 @@ const IntegrationsList: React.FunctionComponent<IntegrationListProps> = ({
+          isModalOpen={isTestModalOpen}
+        />
+      )}
+      {wizardEnabled && category && !modalIsOpenState.isTest && (
```
I'm not a huge fan of making these conditional rendering blocks even more complex than they already are - they do not make it abundantly clear what will render and why (making it harder and harder to diagnose if there is an issue). Also what happens if the delete or create modal are already open in this situation?

florkbr reviewed (commented):
This was merged before I could finish my review. Posting my comments anyway. Seeing styling issues on both stage and locally with the modal (too wide): 
![Screenshot 2023-12-11 at 4 47 44 PM](https://github.com/RedHatInsights/notifications-frontend/assets/21317056/90ae9526-feeb-45d4-9dd1-54a107b06ccf)

It also seems the menu the "test" action appears in can be covered by some of our overlays in the far bottom right corner of the UI (when the "making use of integrations" is open). It's not unique to this PR - though we should investigate a better solution:
![Screenshot 2023-12-11 at 4 47 16 PM](https://github.com/RedHatInsights/notifications-frontend/assets/21317056/8aa3fe6e-a4f4-47b2-b0b3-7afb5e337859)

karelhala commented on src/components/Integrations/Table/IntegrationTest.tsx:
```diff
@@ -0,0 +1,67 @@
+  };
+
+  return (
+    <Modal
```
I think the modal should be 50 % or `medium` or something like that. It feels a bit odd to span across entire screen for just one input. PF has option to change the size, so we should utilize that.
